pkg/snapshot/snapshot.go

```diff
@@ -0,0 +1,141 @@
+package snapshot
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/lightsail"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	defaultSnapshotSuffix = "auto-snapshot"
+)
+
+var (
+	defaultRetention = 10 * 25 * time.Hour
+)
+
+type SnapshotManager struct {
+	client   *lightsail.Lightsail
+	instance string // instance name
+
+	retention time.Duration // retention time
+	suffix    string        // snapshot suffix
+
+	logger log.FieldLogger
+}
+
+type Opt func(*SnapshotManager)
+
+func WithRetention(r time.Duration) Opt {
+	return func(m *SnapshotManager) {
+		m.retention = r
+	}
+}
+
+func WithSnapshotSuffix(suf string) Opt {
+	return func(m *SnapshotManager) {
+		m.suffix = suf
+	}
+}
+
+func NewSnapshotManager(client *lightsail.Lightsail, instance string, opts ...Opt) *SnapshotManager {
+	smgr := &SnapshotManager{
+		client:   client,
+		instance: instance,
+
+		retention: defaultRetention,
+		suffix:    defaultSnapshotSuffix,
+
+		logger: log.New().WithFields(
+			log.Fields{
+				"component": "snapshot-manager",
+				"instance":  instance,
+			}),
+	}
+
+	for _, o := range opts {
+		o(smgr)
+	}
+
+	return smgr
+}
+
+func (smgr *SnapshotManager) Snapshot(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	snapshotName := fmt.Sprintf("%s-%d-%s",
+		smgr.instance,
+		time.Now().UnixNano(),
+		smgr.suffix,
+	)
+	smgr.logger.Infof("Creating snapshot with name %s", snapshotName)
+	// TODO: Check for errors in response
+	_, err := smgr.client.CreateInstanceSnapshotWithContext(
+		ctx,
+		&lightsail.CreateInstanceSnapshotInput{
+			InstanceName:         aws.String(smgr.instance),
+			InstanceSnapshotName: aws.String(snapshotName),
+		},
+	)
+	return err
+}
+
+func (smgr *SnapshotManager) Prune(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
+
+	var snapshots []*lightsail.InstanceSnapshot
+	var token *string
+
+	for {
+		in := &lightsail.GetInstanceSnapshotsInput{}
+		if token != nil {
+			in.PageToken = token
+		}
+		resp, err := smgr.client.GetInstanceSnapshotsWithContext(ctx, in)
+		if err != nil {
+			return err
+		}
+		snapshots = append(snapshots, resp.InstanceSnapshots...)
+		if resp.NextPageToken == nil {
+			break
+		}
+		token = resp.NextPageToken
+	}
+
+	for _, snapshot := range snapshots {
+		if snapshot.CreatedAt == nil {
+			//skip
+			continue
+		}
+		if snapshot.FromInstanceName != nil &&
+			*snapshot.FromInstanceName != smgr.instance {
+			// skip; wrong instance
+			continue
+		}
+
+		// Check if snapshot is old enough was created by this tool
+		if snapshot.CreatedAt.Before(time.Now().Add(-smgr.retention)) &&
```

Comment:
Let's negate the first part of the if-statement and continue with the loop and maybe log that it's skipping a snapshot.

IMO, it's improving the readability.

Reply:
Yeah, I thought about this as well after I had already submitted the PR. Will make some improvements here.